File: mqtt_controller.py
import mqtt_client_helper
import constants
import time
import json
import random

import CowModel
import cow_controler

import SafeZoneModel
import safe_zone_controller

def get_header(msg: str) -> int:
    if len(msg) >= 2: return int(msg[0:2])
    else: return None
    
def handle_mqtt_msg(msg: str):
    """
    This function will handle msg from mqtt
    Params:
        msg (str):
            mqtt message
    """
    # Split message to header and data

    header = get_header(msg)
    
    if(header == constants.HEADER_BACKEND_RESPONSE_GET_COWS):
        cows_data: str = msg[2:]
        data_list = json.loads(cows_data)
        
        for cow_dict in data_list:
            cowModel: CowModel.CowModel = CowModel.CowModel(cow_id=cow_dict['_id'], cow_addr=cow_dict['cow_addr'])
            cow_controler.add_cow(cowModel=cowModel)
        
        print(f"After get all cows: {cow_controler.cows_dict}")
        
    elif(header == constants.HEADER_BACKEND_CREATE_COW):
        # <header><cow>
        # split to get cow_data
        cow_data: str = msg[2:]

        # convert cow_data from string to dict
        cow_dict = json.loads(cow_data)

        # add cow to global variable
        cowModel: CowModel.CowModel = CowModel.CowModel(cow_id=cow_dict['_id'], cow_addr=cow_dict['cow_addr'])
        cow_controler.add_cow(cowModel)

        # response ack to backend
        publish_mqtt_msg(f"0{constants.HEADER_GATEWAY_ACK}{cowModel.cow_id}")

        print(f"After create cow: {cow_controler.cows_dict}")
    
    elif(header == constants.HEADER_BACKEND_DELETE_COW):
        # <header><cow_id>
        # split to get cow id
        cow_id: str = msg[2:]

        #delete cow from global variable
        cow_controler.delete_cow(cow_id=cow_id)
        
        # response ack to backend
        publish_mqtt_msg(f"0{constants.HEADER_GATEWAY_ACK}")
        print(f"After delete cow: {cow_controler.cows_dict}")
    
    elif(header == constants.HEADER_BACKEND_RESPONSE_SAFE_ZONES):
        # <header><safezones>
        # split to get safezones
        safe_zones_data: str = msg[2:]

        # load json
        data_list = json.loads(safe_zones_data)
        # print(data_list)
        for safe_zone_dict in data_list:
            safe_zone_id: str = safe_zone_dict['_id']
            safe_zone_points: list = []

            for point in safe_zone_dict['safeZone']:
                point: SafeZoneModel.Point = SafeZoneModel.Point(point['longitude'], point['latitude'])
                safe_zone_points.append(point)
            
            safeZoneModel: SafeZoneModel.SafeZoneModel = \
                SafeZoneModel.SafeZoneModel(safe_zone_id, safe_zone_points)
            safe_zone_controller.add_safe_zone(safeZoneModel)
        
        print(safe_zone_controller.safe_zones_dict)
        
    else:
        print(f"Header is invalid, header = {header}")
    

def publish_mqtt_msg(msg: str):
    """
    This function will publish `msg` to mqtt broker's `topic` 
    Params:
        msg (str):
            message which we would like to publish
    """
    mqtt_client_helper.mqttClientHelper.publish(msg)


def get_all_cows():
    """
    This function will send a request to get all cows information
to backend
    """

    msg: str = f"0{constants.HEADER_GATEWAY_REQUEST_GET_COWS}" \
                if constants.HEADER_GATEWAY_REQUEST_GET_COWS < 10 \
                else f"{constants.HEADER_GATEWAY_REQUEST_GET_COWS}"
    publish_mqtt_msg(msg)

def send_ack(ack: int):
    """
    This function will feedback ack to backend
    Params:
        ack (int): ack which we wanna feedback to backend
    """

    msg: str = f"0{constants.HEADER_GATEWAY_ACK}{ack}"
    publish_mqtt_msg(msg)

def send_cow_infor(cow_id:str, longitude: float, latitude: float, cow_status):
    msg: str = f"0{constants.HEADER_GATEWAY_SEND_COW_INFOR}" \
                if constants.HEADER_GATEWAY_SEND_COW_INFOR < 10 \
                else f"{constants.HEADER_GATEWAY_SEND_COW_INFOR}"
    msg += f"{cow_id}:{longitude}:{latitude}:{cow_status}"
    publish_mqtt_msg(msg)

def get_safe_zones():
    """
    This function will send a request to get safe zone
    """

    msg: str =  f"0{constants.HEADER_GATEWAY_REQUEST_SAFE_ZONES}" \
                if constants.HEADER_GATEWAY_REQUEST_SAFE_ZONES < 10 \
                else f"{constants.HEADER_GATEWAY_REQUEST_SAFE_ZONES}"
    publish_mqtt_msg(msg)

File: src/controllers/mqttController.js
import admin from "firebase-admin";
import mqtt from "mqtt";

import cowLocationService from "../services/cowLocationService.js";
import cowService from "../services/cowService.js";
import eventService from "../services/eventService.js";
import constants from "./constants.js";
import safeZoneService from "../services/safeZoneService.js";
import mqttService from "../services/mqttService.js";
import serviceAccount from "../config/push-notification-key.json" assert { type: "json" };
import { CowModel } from "../models/cowModel.js";

// Initialize Firebase Admin SDK
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
});

const FCM = admin.messaging();

const handle_mqtt_msg = async (topic, msg) => {
  let data, split_data, cow_id;
  let username, response_msg;
  const header = parseInt(msg.slice(0, 2));
  switch (header) {
    case constants.HEADER_GATEWAY_REQUEST_GET_COWS:
      /* Call service get all cows by username */
      username = topic;
      let cows = await cowService.getCowByUsername(username);
      /* Send response back to gateway */
      response_msg =
        "0" + constants.HEADER_BACKEND_RESPONSE_GET_COWS + JSON.stringify(cows);
      mqttService.publish(topic, response_msg);
      break;
    // case constants.HEADER_GATEWAY_SEND_COW_STATUS:
    //     /* Call service update cow status */
    //     console.log("HEADER_GATEWAY_SEND_COW_STATUS");
    //     data = msg.slice(2);
    //     split_data = data.split(':');
    //     cow_id = split_data[0];
    //     let cow_status = split_data[1];
    //     await cowService.updateCowStatusById(cow_id, cow_status);
    //     break;
    case constants.HEADER_GATEWAY_SEND_COW_INFOR:
      data = msg.slice(2);
      split_data = data.split(":");
      console.log("HEADER_GATEWAY_SEND_COW_INFOR", split_data);
      cow_id = split_data[0];
      let longitude = parseFloat(split_data[1]);
      let latitude = parseFloat(split_data[2]);
      let is_out_safezone = parseInt(split_data[3]);
      /* Call service update latest gps */
      await cowService.updateLatestLocationById(cow_id, longitude, latitude);
      /* Call service create new gps */
      await cowLocationService.createCowLocation(cow_id, longitude, latitude);
      const cow = await CowModel.findById(cow_id);

      if (is_out_safezone && !cow.missing ) {
        /*  Call service to infor out of safezone to user by Firebase */
        cow.missing = true;
        await CowModel.findByIdAndUpdate(
          cow_id, { $set: cow }, { new: true }
        );
        const message = {
          notification: {
            title: "Cattle Out of Safe Zone",
            body: `Cow with ID ${cow_id} has exited the safe zone!`,
          },
          data: {
            type: "Warning",
            cowId: cow_id,
            timestamp: new Date().toISOString(),
          },
          token:
            "cqww-2YiQoGB3KoMJX0AIM:APA91bHxtYJnIUQBFtoXmBbDpnavh4GAezrsG1_Q0kM_6SR9S4-QilURc7ZCDuZn4-tcCyoiC_EYp1-J2ObidfyBuluSvF6VNHKKcGyPUoXmcWkniJ9gous",
        };
        
        FCM.send(message)
          .then((response) => {
            console.log("Notification sent successfully:", response);
          })
          .catch((error) => {
            console.error("Error sending notification:", error);
          });
      }else if(!is_out_safezone && cow.missing){
        cow.missing = false;
        await CowModel.findByIdAndUpdate(
          cow_id, { $set: cow }, { new: true }
        );
      }
      break;
    case constants.HEADER_GATEWAY_REQUEST_SAFE_ZONES:
      /* Call service to get safe zones by username */
      username = topic;
      const safeZone = await safeZoneService.getSafeZoneByUsername(username);
      /* Send response back to gateway */
      response_msg =
        "0" +
        constants.HEADER_BACKEND_RESPONSE_SAFE_ZONES +
        JSON.stringify(safeZone);
      mqttService.publish(topic, response_msg);
      break;
    case constants.HEADER_GATEWAY_ACK:
      /* Emit event */
      eventService.mqttEvent.emit(`${msg}`, "");
      break;

    default:
      // console.log("Header invalid: ", header);
      break;
  }
};
class MQTTClient {
  constructor(username, password, brokerUrl) {
    console.log("init MQTT");
    this.client = mqtt.connect(brokerUrl, {
      username: username,
      password: password,
    });

    this.client.on("connect", () => {
      console.log("Connected to MQTT Broker");
      //subscribe to topic
      this.client.subscribe(`${username}/feeds/+`, (err) => {
        if (err) {
          console.log(err);
        }
      });
    });

    this.client.on("message", async (topic, message) => {
      topic = topic.split("/")[2];
      if (topic == "V1" || topic == "V2") {
        return;
      }

      message = message.toString();
      // console.log(`Received message from ${topic}: ${message}`);
      handle_mqtt_msg(topic, message);
    });

    this.client.on("close", () => {
      console.log("Connection to MQTT Broker closed");
    });
  }
  publish(topic, message) {
    if (!this.client || !this.client.connected) {
      console.error("Client is not initialized or not connected.");
      return;
    }
    // console.log("Publishing to mqtt");
    this.client.publish(`nguyentruongthan/feeds/${topic}`, message);
  }
}
const MQTT_USERNAME = "nguyentruongthan";
const MQTT_PASSWORD = "";
const MQTT_BROKER_URL = "mqtt://mqtt.ohstem.vn";

const mqttClient = new MQTTClient(
  MQTT_USERNAME,
  MQTT_PASSWORD,
  MQTT_BROKER_URL
);
export default mqttClient;
